### DISCUSSION TITLE
Do I need to implement deserialization myself for `Sequential`?

### DISCUSSION THREAD

**Discussion by knota-lab** [tracel-ai/burn#1407] [Q&A] [answered]
```rust
pub struct LoreDetectModel<B: Backend> {
  ...
  wh: SequentialConv2d<B>,
}

#[derive(Module, Debug)]
struct SequentialConv2d<B: Backend> {
    layers: Vec<Conv2d<B>>,
}

impl<B: Backend> SequentialConv2d<B> {
    fn new() -> Self {
        Self { layers: Vec::new() }
    }

    fn from(conv2ds: Vec<Conv2d<B>>) -> Self {
        Self { layers: conv2ds }
    }

    fn add(&mut self, layer: Conv2d<B>) {
        self.layers.push(layer);
    }

    fn forward(&self, x: Tensor<B, 4>) -> Tensor<B, 4> {
        let mut x = x;
        for (index, layer) in self.layers.iter().enumerate() {
            x = layer.forward(x);
            if index < self.layers.len() - 1 {
                x = relu(x);
            }
        }
        x
    }
}
```

```log
called `Result::unwrap()` on an `Err` value: DeserializeError("Serde error: failed to deserialize: Expected struct but got Vec([Map({\"bias\": Map({\"id\": String(\"6676b6aa-4e49-4caa-bb2e-39227ba4e372\"), \"param\": Map({\"shape\": Vec([U64(64)]), \"value\": Vec([F32(0.1547073), F32(-0.016456041), F32(-0.050490323)
```

![image](https://github.com/tracel-ai/burn/assets/26835087/ae13f25c-f9c3-478f-8c9d-b26d7bbd5bcf)

**Comment by antimora** [comment, marked as answer]
Sorry about the bad error message. We will improve it and thanks for filing a bug (https://github.com/tracel-ai/burn/issues/1390). 

No, you do not need to implement any deserialization. You just need to match the keys. If you need to adjust the keys, you can use key remap. Please see this book section: https://burn.dev/book/import/pytorch-model.html#adjusting-the-source-model-architecture.

In your example, you have new `layers` attribute in your model but not in the source. So you'll need to add `layers` before `wh`. Basically, you'll need to replace `model.wh` to `model.wh.layers`.  I recommend this online regex tool that works for Rust regex: https://rregex.dev/?version=1.10&method=replace


Hope this helps! 


P.S. The book online is for Burn 0.12.1 but the main has updated documentation and fixes for the import.
